engine/http.go

```diff
@@ -38,5 +40,20 @@ func (eng *Engine) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	// can only be sent before the body.
 	// (we could possibly use http footers via chunked encoding, but I couldn't find
 	// how to use them in net/http)
+
+	// Cancel the job if the HTTP connection is closed.
+	if closeNotifier, ok := w.(http.CloseNotifier); ok {
```

Comment:
Not sure but looking at the stack trace when builder.run() is called I see this:
```
    /usr/local/go/src/net/http/server.go:1130 +0xbb
    /go/src/github.com/docker/docker/builder/dispatchers.go:241 +0x7c
    /go/src/github.com/docker/docker/builder/evaluator.go:328 +0xa30
    /go/src/github.com/docker/docker/builder/evaluator.go:179 +0x4b8
    /go/src/github.com/docker/docker/builder/job.go:159 +0xdad
    /go/src/github.com/docker/docker/builder/job.go:43 +0x31
    /go/src/github.com/docker/docker/engine/job.go:90 +0x936
    /go/src/github.com/docker/docker/api/server/server.go:1104 +0x12c1
    /go/src/github.com/docker/docker/api/server/server.go:1289 +0x951
    /usr/local/go/src/net/http/server.go:1265 +0x41
    /go/src/github.com/docker/docker/vendor/src/github.com/gorilla/mux/mux.go:98
 +0x2b9
    /usr/local/go/src/net/http/server.go:1703 +0x19a
```
So, if you want to catch this in a generic way then you may need to look at those functions.
But if all you care about is stopping the build() case then let's remove the unused code for now and at least make build() clean, then we can revisit if we can to do this for all operations later (which I think would be a good idea).